### ISSUE TITLE
Add a list index method to the metastore API

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-oss/quickwit#1072]

**Comment by fmassot** [comment]
On this subject, we need to think about how to make it on **file-backed metastore**.

## Solution 1
Create/update a file that contains all the index IDs
Pros:
- just open the file to retrieve the list of index ids
Cons:
- we lose concurrent write on index creation / deletion
- there is one more file to manage

## Solution 2
List the directories and look for `metastore.json` files in each subdirectory
Pros: 
- no need to manage a new file
- no concurrent issues at creation / deletion

Cons:
- a storage engine used for a file-backed metastore needs to support the `list objects` operation. This seems a common API among object storage and also on [IPFS](https://docs.ipfs.io/reference/cli/#ipfs-ls) and HDFS, so it should be ok to add it without limiting Quickwit use cases
- once we got subdirectories name, we can fetch in parallel all the `{subdirectory}/metastore.json` files for each index. 

## Other solution
Do you have anything else in mind @guilload @fulmicoton?

## Your preference

I like solution 2.

**Comment by fmassot** [comment]
## Use cases of FileBackedMetastore

- Testing cases: you want to start fast and test Quickwit on your object storage
- Static cases: you have several indexes on your object storage and want them searchable
- 1 indexer, multiple searchers: in this case, the indexer is updating existing indexes and you want your searchers aware of new splits. 

## Current limitations of the FileBackedMetastore

The file-backed metastore is inherently limited in its usage:
- it cannot support multiple writer nodes
- even with only 1 writer node (indexer), reader nodes (searchers) are not aware of index updates and deletes.

Currently, it's possible to configure a polling parameter so that the metastore fetch regularly the index metadata. Readers can then be aware of metadata updates after a poll. But readers are not currently aware of deletes.

These limitations are acceptable as `FileBackedMetastore` for the use cases listed above.

## Adding a list index to the FileBackedMetastore

### Technical solution: managing a new file that registers indexes states

This file `indexes_states.json` will be the source of truth for the metastore to know which indexes are available.

States `Creating`, `Alive` and `Deleting` are there to distinguish the life cycle of an index:
- `Creating`: this is a transitioning state, the index is being created but we don't know yet if its metadata has been saved
- `Deleting`: this is a transitioning state, the index is being deleted but we don't know yet if its metadata has been deleted
- `Alive`: index is available 

The states are saved in a json file following this schema:
```json
{
  "index_1": "Creating",
  "index_2": "Alive",
  "index_3": "Deleting"
}
```

### Use cases
We want to keep the same use cases as before. 

### Limitations

We will not ensure a full consistency between the list index API and the index metadata you can really fetch. This means that:
-  list index API can return metadata of (index 1, index 2, index 3)
- get index metadata API can return index metadata from index 4

**Comment by fmassot** [comment]
closed by #1169
